zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1165,20 +1177,92 @@ private void reconnect() {
     }
   }
 
+  private class SyncContext {
+    private CountDownLatch _latch;
+    private AtomicInteger _rc;
+
+    public SyncContext(CountDownLatch latch) {
+      _latch = latch;
+    }
+
+    AtomicInteger getRc() {
+      return _rc;
+    }
+
+    void setRc(AtomicInteger rc) {
+      _rc = rc;
+    }
+
+    CountDownLatch getLatch() {
+      return _latch;
+    }
+  }
+
+  private boolean retrySync(String sessionId) throws ZkInterruptedException {
+    if (!_syncOnNewSession) {
+      return true;
+    }
+    while (true) {
```

Comment:
As commented, please check out the other callback implementations. I think the main logic is in common. So we don't need to re-define everything.